timemachine/fe/rbfe.py

```diff
@@ -0,0 +1,378 @@
+    plt.legend()
+
+
+SimulationResult = namedtuple("SimulationResult", ["all_dGs", "all_errs", "plot_png", "frames", "boxes", "U_fns"])
+
+
+def estimate_free_energy_given_samples(all_frames, all_boxes, all_U_fns, lambda_schedule, prefix):
+    """
+    Estimate free energies given pre-generated samples.Let L be the number of lambda windows,
+    T be the number of samples, N be the number of atoms, P be the number of components
+    in the energy function.
+
+    Parameters
+    ----------
+    all_frames: np.array of shape (L, T, N, 3)
+        List of coordinates from all windows
+
+    all_boxes: np.array of shape (L, T, 3, 3)
+        List of boxes from all windows
+
+    all_U_fns: list of energy functions of shape (L, P)
+        List of energy functions from all windows
+
+    lambda_schedule: np.array of shape (L,)
+        Lambda schedule
+
+    prefix: str
+        A prefix to append to overlap figures
+
+    Return
+    ------
+    3-tuple of dGs, dG_errs, overlap_plot
+        All components have have shape (L-1,)
+
+    """
+    # assume pair-BAR format
+    # setup U_fns to operate in batch mode
+    kT = BOLTZ * 300.0
+    beta = 1 / kT
+
+    all_dGs = []
+    all_errs = []
+
+    all_batch_U_fns = []
+    for lamb_idx, lamb in enumerate(lambda_schedule):
+        bound_impls = [p.bound_impl(np.float32) for p in all_U_fns[lamb_idx]]
+        batch_U_fns = get_batch_U_fns(bound_impls, lamb)
+        all_batch_U_fns.append(batch_U_fns)
+
+    U_names = []
+    for U_fn in all_U_fns[0]:
+        # convert from '<timemachine.lib.potentials.Nonbonded object at 0x7f7880b900b8>' -> Nonbonded
+        U_names.append(repr(U_fn).split(".")[-1].split()[0])
+
+    num_rows = len(lambda_schedule) - 1
+    num_cols = len(U_names) + 1
+
+    plot_iota = 1
+
+    def get_plot_idx():
+        nonlocal plot_iota  # what is this black magic? needed to deal with += increment
+        idx = (num_rows, num_cols, plot_iota)
+        plot_iota += 1
+        return idx
+
+    plt.figure(figsize=(num_cols * 5, num_rows * 3))
+    for lamb_idx, lamb in enumerate(lambda_schedule):
+        if lamb_idx > 0:
+
+            prev_frames = all_frames[lamb_idx - 1]
+            cur_frames = all_frames[lamb_idx]
+
+            prev_boxes = all_boxes[lamb_idx - 1]
+            cur_boxes = all_boxes[lamb_idx]
+
+            prev_U_fns = all_batch_U_fns[lamb_idx - 1]
+            cur_U_fns = all_batch_U_fns[lamb_idx]
+
+            # loop over bond, angle, torsion, nonbonded terms etc.
+            all_fwd_delta_us = []
+            all_rev_delta_us = []
+
+            for u_idx, (prev_U_fn, cur_U_fn) in enumerate(zip(prev_U_fns, cur_U_fns)):
+                fwd_delta_u = beta * (cur_U_fn(prev_frames, prev_boxes) - prev_U_fn(prev_frames, prev_boxes))
+                rev_delta_u = beta * (prev_U_fn(cur_frames, cur_boxes) - cur_U_fn(cur_frames, cur_boxes))
+
+                dG, dG_err = pymbar.BAR(fwd_delta_u, rev_delta_u)
+                dG /= beta
```

Comment:
dg /= beta converts to kJ/mol (so update plot_BAR title units or convert after passing to plot_BAR)

Reply:
willfix